app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt

```diff
@@ -0,0 +1,182 @@
+package de.westnordost.streetcomplete.util
+
+import android.util.Log
+import de.westnordost.streetcomplete.data.download.tiles.TilePos
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilePos
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilesRect
+import de.westnordost.streetcomplete.data.download.tiles.minTileRect
+import de.westnordost.streetcomplete.data.osm.mapdata.BoundingBox
+import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
+import de.westnordost.streetcomplete.util.math.contains
+import de.westnordost.streetcomplete.util.math.isCompletelyInside
+
+/**
+ * <Key, LatLon> cache based on tiles, inspired by Android LruCache.
+ * fetch needs get data from db and fill other caches, e.g. questKey -> Quest.
+ * onRemoveTile needs to clean other caches, e.g. questKey -> Quest.
+ * trim() needs to be called after get() and getting data from other caches.
+ * the bbox queried in get() must fit in cache, i.e. may not be larger than maxTiles at tileZoom.
+ */
+class SpatialCache<T>(
```

Comment:
> There seems to be a small discrepancy in documentation comment and code - you mention onRemoveTile but there is not anything like that in the code.

Oh, that should be `onKeysRemoved`...

> unit tests

I did already start on this, but not sure whether I can do this properly.

Reply:
>> unit tests

> I did already start on this, but not sure whether I can do this properly.

What do you need?

Reply:
My initial problem was the conversion between positions in cache and `TilePos`, so I can fill the cache and e.g. put positions/items inside and outside the cached tiles. But I guess this will work if I use positions from the center of the bbox created from `TilePos`

Reply:
> conversion between positions in cache and TilePos

Why, you can use the `LatLon.enclosingTilePos` function for that. That this function works correctly is tested elsewhere.